# SQL Project

## The Goals of the Analysis

The COVID-19 pandemic and the lock downs imposed on people across the world caused people to change their day-to-day
lives. People who used to be busy and barely had time to sit down, were forced to slow down and stay at home. 
This meant that more and more people spent their time reading books - causing an influx of startups developing apps
specifically for book readers and book lovers.

The database analyzed in this analysis is of one of the services competing on this newly, re-emerging, market. This 
database contains information on books, publishers, authors, number of pages, reviewers, reviews and ratings. This information will be used to generate a value proposition for a new product.

## Description of the Tables

1. Books:
    - "book_id" - the book's unique id;
    - "author_id" - the author's unique id;
    - "title" - the book's title;
    - "num_pages" — number of pages;
    - "publication_date" - the book's publication date;
    - "publisher_id" - the publisher's unique id.


2. Authors:
    - "author_id" - the author's unique id;
    - "author" - the author's name.


3. Publishers:
    - "publisher_id" - the publisher's unique id;
    - "publisher" - the publisher's name.


4. Ratings:
    - "rating_id" - the rating's unique id;
    - "book_id" - the book's unique id;
    - "username" — the name of the user who rated the book;
    - "rating" - the book's rating;.


5. Reviews:
    - "review_id" - the review's unique id;;
    - "book_id" - the book's unique id;;
    - "username" — the name of the user who reviewed the book;
    - "text" — the text of the review.

## Loading Libraries and Installing psycopg2

In [1]:
# importing necessary libraries
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth', None)

In [2]:
#!pip install psycopg2

In [3]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Printing the Tables

In [4]:
# creating a loop that prints all the required tables in the database 

tables = ['reviews', 'books', 'publishers', 'ratings', 'authors']
for i in tables:
    print("The {} table:".format(i))
    query = '''SELECT *
         FROM {}
         LIMIT 10'''.format(i)

    display(pd.read_sql_query(query, con = engine))

The reviews table:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.  
5                                                                                          Analysis no several cause international.  
6                              One there cost another. Say type save. With personal where occur direction foot. Half sit role want.  
7                                                                 Within enough mother. There at system full recent play recognize.  
8                                  Thank now focus realize economy focus fly. Item step fine war. Western service history anything.  
9            Game push lot reduce where remember. Including song travel hair foot. Director always myself past story chance reason.

The books table:


book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   
5        6        257                   1st to Die (Women's Murder Club  #1)   
6        7        258                   2nd Chance (Women's Murder Club  #2)   
7        8        260                  4th of July (Women's Murder Club  #4)   
8        9        563                                       A Beautiful Mind   
9       10        445                                     A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

The publishers table:


publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

The ratings table:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

The authors table:


author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

## Analysis of the Data

### Number of Books Released After January 1st, 2020

In [5]:
query = '''SELECT COUNT(title)
           FROM books
           WHERE publication_date > CAST('2000-01-01' as Date);'''

display(pd.read_sql_query(query, con = engine))
      

count
0    819

### Number of Reviews and Average Ratings per Book

In [6]:
query = '''SELECT books.book_id,
           books.title, 
           COUNT(DISTINCT reviews.review_id) AS n_reviews,
           AVG(ratings.rating) AS avg_rating
           FROM books
           LEFT JOIN ratings ON books.book_id = ratings.book_id
           LEFT JOIN reviews ON ratings.book_id = reviews.book_id
           GROUP BY books.book_id
           ORDER BY n_reviews DESC;'''

display(pd.read_sql_query(query, con = engine))



book_id                                                        title  \
0        948                                      Twilight (Twilight  #1)   
1        963                                          Water for Elephants   
2        734                                             The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)   
4        695            The Curious Incident of the Dog in the Night-Time   
..       ...                                                          ...   
995       83              Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                                      The Natural Way to Draw   
997      672             The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                                    Essential Tales and Poems   
999      191         Disney's Beauty and the Beast (A Little Golden Book)   

     n_reviews  avg_rating  
0            7    3.662500  
1            6    3.977273  
2            6    4.206897  
3            6    4.414634  
4            6    4.081081  
..         ...         ...  
995          0    3.666667  
996          0    3.000000  
997          0    5.000000  
998          0    4.000000  
999          0    4.000000  

[1000 rows x 4 columns]

### Publisher with the Greatest Number of Books Released (with More than 50 Pages)

In [9]:
query = '''SELECT publishers.publisher_id,
           publishers.publisher,
           COUNT(books.publisher_id) AS n_books
           FROM publishers
           LEFT JOIN books ON publishers.publisher_id = books.publisher_id
           WHERE num_pages > 50
           GROUP BY publishers.publisher_id
           ORDER BY n_books DESC
           LIMIT 1;'''


display(pd.read_sql_query(query, con = engine))

publisher_id      publisher  n_books
0           212  Penguin Books       42

### Author with the Highest Average Book Ratings (At Least 50 Ratings)


In [10]:
query = '''SELECT top.author,
           top.avg_rating,
           top.n_ratings
           FROM (
           SELECT books.title,
           authors.author,
           AVG(ratings.rating) AS avg_rating,
           COUNT(ratings.rating) AS n_ratings
           FROM authors
           LEFT JOIN books ON books.author_id = authors.author_id
           LEFT JOIN ratings ON ratings.book_id = books.book_id
           GROUP BY books.title, authors.author
           HAVING COUNT(ratings.rating) >= 50
           ORDER BY avg_rating DESC
           LIMIT 1) AS top
           '''

display(pd.read_sql_query(query, con = engine))

author  avg_rating  n_ratings
0  J.K. Rowling/Mary GrandPré    4.414634         82

### Average Number of Text Reviews (Users with More Than 50 Books Rated)

In [11]:
query = '''WITH
           i AS (SELECT reviews.username,
           COUNT(reviews.review_id) AS count
           FROM reviews
           WHERE username in (SELECT username
           FROM ratings
           GROUP BY username
           HAVING COUNT(ratings.username) > 50)
           GROUP BY reviews.username)
           
           SELECT AVG(i.count) AS avg_reviews
           FROM i
           '''


display(pd.read_sql_query(query, con = engine))

avg_reviews
0    24.333333

# Conclusions

Conclusions on each of the queries above: 
- Out of 1,000 books, more than 80% were published after January 1st, 2020. This means that newer books sell better than older ones (even better than the classics).
- Top five include both books that are part of series and some more classic books. However, all were most likely released more than 10 years ago, at the very least. It seems that people were nostalgic during the pandemic - rather than chasing trends.
- Some publishing houses are dominating the publishing world - Penguin has the most books published (42), then Vintage (which is also part of Penguin Books) and Penguin Classics at 4th. 
- J.K. Rowling has the highest average rating for a book with more than 50 reviews. 
- Readers who rated more than 50 books have an average of 24.33 reviews. This means that not everyone who rates a book, leaves a text review of the book or a book.